File: cue/core/voice/src/main/kotlin/dev/cue/voice/Text.kt
package dev.cue.voice

/**
 * The shared text primitives. Everything in §4 and §7 is built on these, so
 * they are defined once and tested directly rather than re-derived per call
 * site with slightly different regexes.
 */
object Text {
    /**
     * A word is letters and digits, optionally carrying an internal apostrophe
     * so "don't" and "i'm" survive tokenisation intact — the contraction rate
     * (§4.1) is meaningless if the tokeniser splits them.
     */
    private val WORD = Regex("""[\p{L}\p{N}]+(?:['’][\p{L}]+)*""")

    /** Sentence-ending punctuation as it appears in messages, ellipsis included. */
    val TERMINAL_PUNCTUATION = charArrayOf('.', '!', '?', '…')

    fun words(text: String): List<String> = WORD.findAll(text).map { it.value }.toList()

    fun wordCount(text: String): Int = WORD.findAll(text).count()

    /** Lowercased and stripped of curly apostrophes, for comparison and counting. */
    fun normalize(word: String): String = word.lowercase().replace('’', '\'')

    fun normalizedWords(text: String): List<String> = words(text).map(::normalize)

    /** The first alphabetic character, skipping leading emoji, quotes and spaces. */
    fun firstLetter(text: String): Char? = text.firstOrNull { it.isLetter() }

    /** True when the message ends in terminal punctuation, ignoring trailing emoji. */
    fun endsWithTerminalPunctuation(text: String): Boolean {
        val last = trimTrailingNonPunctuation(text).lastOrNull() ?: return false
        return last in TERMINAL_PUNCTUATION
    }

    /**
     * True when the message is a question, ignoring trailing emoji.
     *
     * Shares [trimTrailingNonPunctuation] with [endsWithTerminalPunctuation] on
     * purpose. When the two disagreed — a plain `endsWith('?')` here and an
     * emoji-tolerant check there — "how was your week? 😂" counted as a
     * statement *and* as punctuated, which is the only combination that cannot
     * be true, and it put a lowercase writer's statement-punctuation rate above
     * zero.
     */
    fun endsWithQuestionMark(text: String): Boolean =
        trimTrailingNonPunctuation(text).endsWith('?')

    /**
     * Walks backwards by code point, not by `Char`.
     *
     * Every modern emoji is a surrogate pair, so a `Char`-wise scan sees a low
     * surrogate that matches no emoji range, stops there, and concludes that
     * "that's the plan. 😂" does not end in punctuation. The consequence is a
     * `terminalPunctuationRate` measured on the wrong messages.
     */
    private fun trimTrailingNonPunctuation(text: String): String {
        var end = text.length
        while (end > 0) {
            val cp = text.codePointBefore(end)
            val skippable = Character.isWhitespace(cp) ||
                Emoji.isEmojiCodePoint(cp) ||
                cp in EMOJI_MODIFIERS ||
                cp == '"'.code ||
                cp == '\''.code
            if (!skippable) break
            end -= Character.charCount(cp)
        }
        return text.substring(0, end)
    }

    /** Variation selectors, ZWJ and skin tones — trailing, never terminal. */
    private val EMOJI_MODIFIERS = setOf(0xFE0F, 0xFE0E, 0x200D, 0x20E3) +
        (0x1F3FB..0x1F3FF).toSet()

    fun countChar(text: String, c: Char): Int = text.count { it == c }

    /**
     * Occurrences of the standalone first-person pronoun, in both cases.
     *
     * Standalone means not part of a longer word — "in" and "island" must not
     * count — but "I'm" and "I'd" do, because they are the same pronoun and the
     * capitalisation decision is identical.
     */
    private val STANDALONE_I = Regex("""(?<![\p{L}])([Ii])(?=(?:['’][\p{L}]+)?(?![\p{L}]))""")

    fun standaloneIOccurrences(text: String): List<MatchResult> =
        STANDALONE_I.findAll(text).toList()

    /**
     * Splits into clauses at punctuation and coordinating conjunctions, keeping
     * the boundary offsets. §4.4's length transform truncates *at* a boundary
     * rather than mid-phrase, which is the difference between a short message
     * and a message that got cut off.
     */
    fun clauseBoundaries(text: String): List<Int> {
        val boundaries = mutableListOf<Int>()
        var i = 0
        while (i < text.length) {
            val c = text[i]
            if (c in TERMINAL_PUNCTUATION || c == ',' || c == ';') {
                // Consume a run, so "..." and "?!" yield one boundary.
                var end = i + 1
                while (end < text.length && (text[end] in TERMINAL_PUNCTUATION || text[end] == ',' || text[end] == ';')) {
                    end++
                }
                boundaries += end
                i = end
            } else {
                i++
            }
        }
        for (conjunction in CONJUNCTIONS) {
            var from = 0
            while (true) {
                val at = text.indexOf(conjunction, from, ignoreCase = true)
                if (at < 0) break
                boundaries += at
                from = at + 1
            }
        }
        return boundaries.filter { it in 1 until text.length }.distinct().sorted()
    }

    private val CONJUNCTIONS = listOf(" and ", " but ", " so ", " because ", " though ", " while ")

    /** Collapses runs of whitespace and trims. Applied before anything measures length. */
    fun tidy(text: String): String = text.replace(Regex("""[ \t]+"""), " ")
        .replace(Regex(""" ?\n ?"""), "\n")
        .trim()
}

/**
 * Emoji scanning by code point.
 *
 * Kotlin has no emoji API, and treating each `Char` as one emoji is wrong for
 * everything above the BMP — which is most emoji. §4.1's `emojiRate` and §4.4's
 * trim both need whole graphemes: a family emoji is one emoji, not four, and
 * removing "the third char" of one produces mojibake in a message you are about
 * to send someone.
 */
object Emoji {
    private const val ZWJ = 0x200D
    private const val VS16 = 0xFE0F
    private const val VS15 = 0xFE0E
    private const val KEYCAP = 0x20E3
    private val SKIN_TONES = 0x1F3FB..0x1F3FF
    private val REGIONAL_INDICATORS = 0x1F1E6..0x1F1FF

    private val BASE_RANGES = listOf(
        0x1F000..0x1F02F, // mahjong, dominoes
        0x1F0A0..0x1F0FF, // playing cards
        0x1F300..0x1FAFF, // the bulk of modern emoji
        0x2600..0x27BF, // misc symbols and dingbats
        0x2B00..0x2BFF, // stars, arrows
        0x2190..0x21FF, // arrows (only ever emoji with VS16, but harmless here)
        0x2900..0x297F,
        0x3030..0x3030,
        0x303D..0x303D,
        0x3297..0x3299,
    )

    fun isEmojiCodePoint(cp: Int): Boolean =
        cp in REGIONAL_INDICATORS || BASE_RANGES.any { cp in it }

    /**
     * Returns each emoji's character range, with ZWJ sequences, skin-tone
     * modifiers, keycaps, flags and variation selectors folded into one entry.
     */
    fun ranges(text: String): List<IntRange> {
        val out = mutableListOf<IntRange>()
        var i = 0
        while (i < text.length) {
            val cp = text.codePointAt(i)
            val width = Character.charCount(cp)
            if (!isEmojiCodePoint(cp)) {
                // A lone digit or '#' followed by a keycap is one emoji.
                if (i + width < text.length && text.codePointAt(i + width) == VS16 &&
                    i + width + 1 < text.length && text.codePointAt(i + width + 1) == KEYCAP
                ) {
                    out += i..(i + width + 1)
                    i += width + 2
                    continue
                }
                i += width
                continue
            }
            var end = i + width
            if (cp in REGIONAL_INDICATORS && end < text.length) {
                val next = text.codePointAt(end)
                if (next in REGIONAL_INDICATORS) end += Character.charCount(next)
            }
            while (end < text.length) {
                val next = text.codePointAt(end)
                val nextWidth = Character.charCount(next)
                when {
                    next == VS16 || next == VS15 || next == KEYCAP -> end += nextWidth
                    next in SKIN_TONES -> end += nextWidth
                    next == ZWJ -> {
                        val after = end + nextWidth
                        if (after >= text.length) break
                        end = after + Character.charCount(text.codePointAt(after))
                    }
                    else -> break
                }
            }
            out += i until end
            i = end
        }
        return out
    }

    fun findAll(text: String): List<String> = ranges(text).map { text.substring(it.first, it.last + 1) }

    fun count(text: String): Int = ranges(text).size
}

File: cue/extension/src/lib/text.js
// The text primitives, ported from :core:voice's Text.kt.
//
// This file is a port and not a shared library, which is a real cost: two
// implementations of §4.4 can drift. The alternative was a build step to compile
// Kotlin/JS, and §3.2 keeps the whole dependency list at "free, offline, and
// unlimited" — adding a toolchain to save 300 lines of JavaScript is the wrong
// trade for a project whose §15 milestone 7 is "parity with mobile".
//
// What keeps them honest is `test/voice.test.js`, which asserts the same
// behaviours as `VoiceCompilerTest.kt` against the same inputs. When the Kotlin
// changes, that file is where the drift shows up.

const WORD_PATTERN = /[\p{L}\p{N}]+(?:['’]\p{L}+)*/gu;

export const TERMINAL_PUNCTUATION = ['.', '!', '?', '…'];

export function words(text) {
  return Array.from(text.matchAll(WORD_PATTERN), (match) => match[0]);
}

export function wordCount(text) {
  return words(text).length;
}

export function normalize(word) {
  return word.toLowerCase().replace(/’/g, "'");
}

export function normalizedWords(text) {
  return words(text).map(normalize);
}

export function countChar(text, character) {
  let count = 0;
  for (const char of text) if (char === character) count += 1;
  return count;
}

export function firstLetterIndex(text) {
  const match = /\p{L}/u.exec(text);
  return match ? match.index : -1;
}

// Emoji, by code point. JavaScript strings are UTF-16 like Kotlin's, so the same
// trap applies: iterating by index splits every astral emoji in half.
const EMOJI_RANGES = [
  [0x1f000, 0x1f02f],
  [0x1f0a0, 0x1f0ff],
  [0x1f300, 0x1faff],
  [0x2600, 0x27bf],
  [0x2b00, 0x2bff],
  [0x2190, 0x21ff],
  [0x2900, 0x297f],
  [0x3030, 0x3030],
  [0x303d, 0x303d],
  [0x3297, 0x3299],
];
const REGIONAL = [0x1f1e6, 0x1f1ff];
const ZWJ = 0x200d;
const VARIATION = [0xfe0f, 0xfe0e, 0x20e3];
const SKIN_TONES = [0x1f3fb, 0x1f3ff];

export function isEmojiCodePoint(code) {
  if (code >= REGIONAL[0] && code <= REGIONAL[1]) return true;
  return EMOJI_RANGES.some(([low, high]) => code >= low && code <= high);
}

/** Whole emoji graphemes, with ZWJ sequences and modifiers folded in. */
export function emojiRanges(text) {
  const found = [];
  let index = 0;
  while (index < text.length) {
    const code = text.codePointAt(index);
    const width = code > 0xffff ? 2 : 1;
    if (!isEmojiCodePoint(code)) {
      index += width;
      continue;
    }
    let end = index + width;
    if (code >= REGIONAL[0] && code <= REGIONAL[1] && end < text.length) {
      const next = text.codePointAt(end);
      if (next >= REGIONAL[0] && next <= REGIONAL[1]) end += next > 0xffff ? 2 : 1;
    }
    while (end < text.length) {
      const next = text.codePointAt(end);
      const nextWidth = next > 0xffff ? 2 : 1;
      if (VARIATION.includes(next) || (next >= SKIN_TONES[0] && next <= SKIN_TONES[1])) {
        end += nextWidth;
      } else if (next === ZWJ) {
        const after = end + nextWidth;
        if (after >= text.length) break;
        const joined = text.codePointAt(after);
        end = after + (joined > 0xffff ? 2 : 1);
      } else {
        break;
      }
    }
    found.push([index, end]);
    index = end;
  }
  return found;
}

export function emojiList(text) {
  return emojiRanges(text).map(([start, end]) => text.slice(start, end));
}

export function emojiCount(text) {
  return emojiRanges(text).length;
}

// `Array.from` on a string yields whole code points, which is the whole trick:
// indexing by `.length` walks into the middle of an emoji and concludes that
// "that's the plan. 😂" does not end in punctuation.
function trimTrailingNonPunctuation(text) {
  const points = Array.from(text);
  let end = points.length;
  while (end > 0) {
    const point = points[end - 1].codePointAt(0);
    const skippable =
      /\s/u.test(points[end - 1]) ||
      isEmojiCodePoint(point) ||
      VARIATION.includes(point) ||
      point === ZWJ ||
      (point >= SKIN_TONES[0] && point <= SKIN_TONES[1]) ||
      points[end - 1] === '"' ||
      points[end - 1] === "'";
    if (!skippable) break;
    end -= 1;
  }
  return points.slice(0, end).join('');
}

export function endsWithTerminalPunctuation(text) {
  const trimmed = trimTrailingNonPunctuation(text);
  return TERMINAL_PUNCTUATION.includes(trimmed.slice(-1));
}

export function endsWithQuestionMark(text) {
  return trimTrailingNonPunctuation(text).endsWith('?');
}

const CONJUNCTIONS = [' and ', ' but ', ' so ', ' because ', ' though ', ' while '];

/** Offsets a message can be cut at without reading as interrupted. */
export function clauseBoundaries(text) {
  const boundaries = new Set();
  let index = 0;
  while (index < text.length) {
    const char = text[index];
    if (TERMINAL_PUNCTUATION.includes(char) || char === ',' || char === ';') {
      let end = index + 1;
      while (
        end < text.length &&
        (TERMINAL_PUNCTUATION.includes(text[end]) || text[end] === ',' || text[end] === ';')
      ) {
        end += 1;
      }
      boundaries.add(end);
      index = end;
    } else {
      index += 1;
    }
  }
  for (const conjunction of CONJUNCTIONS) {
    let from = 0;
    for (;;) {
      const at = text.toLowerCase().indexOf(conjunction, from);
      if (at < 0) break;
      boundaries.add(at);
      from = at + 1;
    }
  }
  return Array.from(boundaries)
    .filter((offset) => offset >= 1 && offset < text.length)
    .sort((a, b) => a - b);
}

export function tidy(text) {
  return text
    .replace(/[ \t]+/g, ' ')
    .replace(/ ?\n ?/g, '\n')
    .trim();
}
